File: 30 November 2022/Python/Reorder List.py
#User function Template for python3

#User function Template for python3

# Node Class    
class node:
    def __init__(self, val):
        self.data = val
        self.next = None

# Linked List Class
class Linked_List:
    def __init__(self):
        self.head = None
        self.tail = None


#Back-end complete function Template for Python 3
class Solution:
    
    def reorderList(self,head):
        if (head==None or head.next==None):
            return
        n=0
        walk=head
        while(walk):
            walk=walk.next
            n=n+1
        head2=head
        for i in range((n-1)//2):
            head2=head2.next
        tail=head2
        head2=head2.next
        tail.next=None
        prv=None
        nex=head2.next
        while(head2):
            nex=head2.next
            head2.next=prv
            prv=head2
            head2=nex
        head2=prv
        walk=head
        walk2=head2
        while(walk2):
            temp=walk.next
            walk.next=walk2
            walk=temp
            temp=walk2.next
            walk2.next=walk
            walk2=temp
        if( tail.next ):
            tail=tail.next

#{ 
 # Driver Code Starts
#Initial Template for Python 3

# Node Class    
class node:
    def __init__(self, val):
        self.data = val
        self.next = None

# Linked List Class
class Linked_List:
    def __init__(self):
        self.head = None
        self.tail = None

    def insert(self, val):
        if self.head == None:
            self.head = node(val)
            self.tail = self.head
        else:
            new_node = node(val)
            self.tail.next = new_node
            self.tail = new_node

    def createList(self, arr, n):
        for i in range(n):
            self.insert(arr[i])
        return self.head

    def printList(self,head):
        tmp = head
        while tmp is not None:
            print(tmp.data, end=" ")
            tmp=tmp.next
        print()

if __name__=='__main__':
    t = int(input())
    for i in range(t):
        n = int(input())
        arr = list(map(int, input().strip().split()))
        lis = Linked_List()
        head = lis.createList(arr, n)
        ob=Solution()
        ob.reorderList(head)

        lis.printList(head)

# } Driver Code Ends

File: Problems & Solutions/R/Reorder List/30 November 2022/C++/Reorder List.cpp
//{ Driver Code Starts
#include <stdio.h>
#include <stdlib.h>
#include <iostream>
#include <bits/stdc++.h>
using namespace std;
/* Linked list Node */
struct Node {
    int data;
    struct Node* next;

    Node(int x) {
        data = x;
        next = NULL;
    }
};



// } Driver Code Ends
/* Following is the Linked list node structure */

/*
struct Node
{
    int data;
    struct Node* next;

    Node(int x){
        data = x;
        next = NULL;
    }
};
*/

class Solution{ 
public:
    
    void reorderList(struct Node* head) {
        int r = 0, c;
        struct Node* temp, *temp1, *temp2, *p;
        temp = head;

        if (head == NULL || head->next == NULL || head->next->next == NULL) return;

        while (temp != NULL) {
            r++;
            temp = temp->next;
        }

        c = (r + 1) / 2;

        temp = head;
        while (c--) {
            p = temp;
            temp = temp->next;
        }

        p->next = NULL;

        temp1 = NULL;
        temp2 = temp;

        while (temp2 != NULL) {
            p = temp2->next;
            temp2->next = temp1;
            temp1 = temp2;
            temp2 = p;
        }

        p = temp1;
        Node *a, *b;
        while (p) {
           a = head->next;
           b = p->next;
           
           head->next=p;
           p->next=a;
           
           head=a;
           p=b;
        }
    }
};

//{ Driver Code Starts.



/* Function to create a new Node with given data */
struct Node* newNode(int data) {
    struct Node* new_Node = new Node(data);
    new_Node->data = data;
    new_Node->next = NULL;
    return new_Node;
}

void printList(struct Node* Node) {
    while (Node != NULL) {
        printf("%d ", Node->data);
        Node = Node->next;
    }
    printf("\n");
}

void freeList(struct Node* head) {
    struct Node* temp;
    while (head != NULL) {

        temp = head;
        head = head->next;
        free(temp);
    }
}

int main(void) {
    int t, n, m, i, x;

    cin >> t;
    while (t--) {
        struct Node* temp, *head;
        cin >> n;
        cin >> x;
        head = new Node(x);
        temp = head;
        for (i = 0; i < n - 1; i++) {
            cin >> x;
            temp->next = new Node(x);
            temp = temp->next;
        }
        
        Solution ob;
        
        ob.reorderList(head);
        printList(head);
        freeList(head);
    }
    return 0;
}

// } Driver Code Ends